x-pack/solutions/observability/plugins/observability/server/services/related_dashboards_client.ts

```diff
@@ -0,0 +1,253 @@
+import type { SavedObjectsFindResult } from '@kbn/core/server';
+import { IContentClient } from '@kbn/content-management-plugin/server/types';
+import type {
+  FieldBasedIndexPatternColumn,
+  GenericIndexPatternColumn,
+} from '@kbn/lens-plugin/public';
+import type { Logger } from '@kbn/core/server';
+import type { LensAttributes } from '@kbn/lens-embeddable-utils';
+import type { RelevantPanel, RelatedDashboard } from '@kbn/observability-schema';
+import type { DashboardAttributes, DashboardPanel } from '@kbn/dashboard-plugin/server';
+import type { InvestigateAlertsClient } from './investigate_alerts_client';
+import type { AlertData } from './alert_data';
+
+type Dashboard = SavedObjectsFindResult<DashboardAttributes>;
+export class RelatedDashboardsClient {
+  private dashboardsById = new Map<string, Dashboard>();
+  private alert: AlertData | null = null;
+
+  constructor(
+    private logger: Logger,
+    private dashboardClient: IContentClient<Dashboard>,
+    private alertsClient: InvestigateAlertsClient,
+    private alertId: string
+  ) {}
+
+  async fetchSuggestedDashboards(): Promise<{ suggestedDashboards: RelatedDashboard[] }> {
+    const allRelatedDashboards = new Set<RelatedDashboard>();
+    const relevantDashboardsById = new Map<string, RelatedDashboard>();
+    const [alert] = await Promise.all([
+      this.alertsClient.getAlertById(this.alertId),
+      this.fetchDashboards(),
+    ]);
+    this.alert = alert;
+    if (!this.alert) {
+      return { suggestedDashboards: [] };
+    }
+    const index = await this.getRuleQueryIndex();
+    const relevantRuleFields = this.alert.getRelevantRuleFields();
+    const relevantAlertFields = this.alert.getRelevantAADFields();
+    const allRelevantFields = new Set([...relevantRuleFields, ...relevantAlertFields]);
+
+    if (index) {
+      const { dashboards } = this.getDashboardsByIndex(index);
+      dashboards.forEach((dashboard) => allRelatedDashboards.add(dashboard));
+    }
+    if (allRelevantFields.size > 0) {
+      const { dashboards } = this.getDashboardsByField(Array.from(allRelevantFields));
+      dashboards.forEach((dashboard) => allRelatedDashboards.add(dashboard));
+    }
+    allRelatedDashboards.forEach((dashboard) => {
+      const dedupedPanels = this.dedupePanels([
+        ...(relevantDashboardsById.get(dashboard.id)?.relevantPanels || []),
+        ...dashboard.relevantPanels,
+      ]);
+      const relevantPanelCount = dedupedPanels.length;
+      relevantDashboardsById.set(dashboard.id, {
+        ...dashboard,
+        matchedBy: {
+          ...relevantDashboardsById.get(dashboard.id)?.matchedBy,
+          ...dashboard.matchedBy,
+        },
+        relevantPanelCount,
+        relevantPanels: dedupedPanels,
+      });
+    });
+    return { suggestedDashboards: Array.from(relevantDashboardsById.values()) };
+  }
+
+  async fetchDashboards() {
+    const dashboards = await this.dashboardClient.search({}, { spaces: ['*'] });
```

Comment:
This will return a paginated response with a default of 20 items per page. Is that what you are expecting?

Reply:
I updated the pagination logic. Feel free to take a look. Thanks a bunch!